python/lsst/ap/association/diaCalculation.py

```diff
@@ -0,0 +1,291 @@
+        created.
+    **kwargs
+        Additional arguments passed to the superclass constructor.
+
+    Notes
+    -----
+    Plugins may either take an entire catalog to work on at a time, or work on
+    individual records.
+    """
+    ConfigClass = DiaObjectCalculationConfig
+    _DefaultName = "diaObjectCalculation"
+
+    def __init__(self, plugMetadata=None, **kwargs):
+        lsst.pipe.base.Task.__init__(self, **kwargs)
+        if plugMetadata is None:
+            plugMetadata = lsst.daf.base.PropertyList()
+        self.plugMetadata = plugMetadata
+        self.plugins = PluginMap()
+
+        self.initializePlugins()
+
+    def initializePlugins(self):
+        """Initialize the plugins according to the configuration.
+        """
+
+        pluginType = namedtuple('pluginType', 'single multi')
+        self.executionDict = {}
+        # Read the properties for each plugin. Allocate a dictionary entry for
+        # each run level. Verify that the plugins are above the minimum run
+        # level for an catalogCalculation plugin. For each run level, the
+        # plugins are sorted into either single record, or multi record groups
+        # to later be run appropriately
+        for executionOrder, name, config, PluginClass in sorted(self.config.plugins.apply()):
+            if executionOrder not in self.executionDict:
+                self.executionDict[executionOrder] = pluginType(single=[], multi=[])
+            if PluginClass.getExecutionOrder() >= BasePlugin.DEFAULT_CATALOGCALCULATION:
+                plug = PluginClass(config, name, metadata=self.plugMetadata)
+                self.plugins[name] = plug
+                if plug.plugType == 'single':
+                    self.executionDict[executionOrder].single.append(plug)
+                elif plug.plugType == 'multi':
+                    self.executionDict[executionOrder].multi.append(plug)
+            else:
+                errorTuple = (PluginClass, PluginClass.getExecutionOrder(),
+                              BasePlugin.DEFAULT_CATALOGCALCULATION)
+                raise ValueError("{} has an execution order less than the minimum for an catalogCalculation "
+                                 "plugin. Value {} : Minimum {}".format(*errorTuple))
+
+    @lsst.pipe.base.timeMethod
+    def run(self, diaObjectCat, diaSourceCat, updatedDiaObjectIds, filterName):
+        """The entry point for the DIA catalog calculation task.
+
+        Run method both updates the values in the diaObjectCat and appends
+        newly created DiaObjects to the catalog.
+
+        Parameters
+        ----------
+        diaObjectCat : `pandas.DataFrame`
+            DiaObjects to update values of and append new objects to. DataFrame
+            should be indexed on "diaObjectId"
+        diaSourceCat : `pandas.DataFrame`
+            DiaSources associated with the DiaObjects in diaObjectCat. DataFrame
+            should be indexed on `["diaObjectId", "filterName", "diaSourceId"]`
```

Comment:
Crash. I assume I should tests for the correct indexing and throw my own error if it is not present?

Reply:
I think it depends on the error that is being thrown at present. If calling it with some incorrectly-indexed DataFrame will actually throw a sensible exception, that's fine and I don't think you need add any more logic here. However, I would ask that you update the docstring to:

- Say “must be indexed” rather than “should be...”
- Add a (`Raises` block)[https://developer.lsst.io/python/numpydoc.html#py-docstring-raises] explaining the exception which will be thrown in the event of improper indexing.

Reply:
Added. For reference pandas will raise KeyError